docs/source/model_library.rst

```diff
@@ -0,0 +1,550 @@
+
+.. code-block:: pycon
+
+   >>> library = ModelLibrary("my_asn.json")
+
+.. NOTE::
+   For an association that contains a ``group_id`` for each member,
+   creating a library will not read any models.
+
+.. _library_borrowing_and_shelving:
+
+Borrowing and shelving models
+-----------------------------
+
+Interacting with an `~stpipe.library.AbstractModelLibrary` involves "borrowing" and "shelving"
+models, both of which must occur during a ``with`` statement (while the library
+is "open"):
+
+.. code-block:: pycon
+
+   >>> with library:
+   ...    model = library.borrow(0)
+   ...    # do stuff with the model...
+   ...    library.shelve(model)
+
+Iteration is also supported (but don't forget to return your models!).
+
+.. code-block:: pycon
+
+   >>> with library:
+   ...    for model in library:
+   ...        # do stuff with the model...
+   ...        library.shelve(model)
+
+
+.. _library_on_disk:
+
+On Disk Mode
+------------
+
+For large associations (like those larger than memory) it is important
+that the library avoid reading all models at once. The borrow/shelve API
+above maps closely to the loading/saving of input (or tempoary) files
+containing the models.
+
+.. code-block:: pycon
+
+   >>> library = ModelLibrary("my_big_asn.json", on_disk=True)
+   >>> with library:
+   ...     model = library.borrow(0)  # the input file for model 0 is loaded
+   ...     library.shelve(model)  # a temporary file for model 0 is written
+
+.. NOTE::
+   In the above example, a temporary file was created for model 0. At no
+   point will the library overwrite the input file.
+
+If model is not modified during the time it's borrowed (for example if the
+``model.dq`` array was read, but not modified). It is helpful to tell the
+library that the model was not modified.
+
+.. code-block:: pycon
+
+   >>> with library:
+   ...     model = library.borrow(0)  # the input file for model 0 is loaded
+   ...     # do some read-only stuff with the model
+   ...     library.shelve(model, modify=False)  # No temporary file will be written
+
+This can dramatically reduce the number of times a file is written saving
+on both disk space and the time required to write.
+
+
+.. _library_map_function:
+
+Map function
+------------
+
+Let's say you want to get the ``meta.filename`` attribute for all models
+in a library. The above "open", "borrow", "shelve" pattern can be quite
+verbose. Instead, the helper method `~stpipe.library.AbstractModelLibrary.map_function`
+can be used to generate an iterator that returns the result of a function
+applied to each model in the library:
+
+.. code-block:: pycon
+
+   >>> def get_model_name(model, index):
+   ...     return model.meta.filename
+   >>>
+   >>> filenames = list(library.map_function(get_model_name))
+
+.. NOTE::
+   `~stpipe.library.AbstractModelLibrary.map_function` does not require an open library
+   and will handle opening, borrowing, shelving and closing for you.
+
+
+.. _library_grouping:
+
+Grouping
```

Comment:
it occurs to me at this point that it might be good to have an additional section (perhaps very near the top) that just discusses how to access the association metadata, i.e., what the names of all the metadata attributes are and what they return.

Reply:
I expanded this a bit in 55ccf14

I don't think we can describe all the association metadata here (as there are differences between roman and jwst). I attempted to describe the common structure required by `ModelLibrary`.

Reply:
yeah I figured that would be the case.  What you did is already super helpful.  I can try to fill in the gaps when documenting the JWST side